lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -2206,15 +2206,36 @@ LLVMToSPIRVBase::transValueWithoutDecoration(Value *V, SPIRVBasicBlock *BB,
   if (AllocaInst *Alc = dyn_cast<AllocaInst>(V)) {
     SPIRVType *TranslatedTy = transScavengedType(V);
     if (Alc->isArrayAllocation()) {
+      SPIRVValue *Length = transValue(Alc->getArraySize(), BB);
+      assert(Length && "Couldn't translate array size!");
+
+      if (isSpecConstantOpCode(Length->getOpCode())) {
+        // SPIR-V arrays length can be expressed using a specialization
+        // constant.
+        //
+        // Spec Constant Length Arrays need special treatment, as the allocation
```

Comment:
I must admit, but I don't understand that. In LLVM IR in the test I see a 1-dimentional array, while here a 2-dimentional array is being created in SPIR-V. Is there some validation rule I'm missing?

Reply:
What this patch intends to do is to represent `alloca <type>, <ty> %specialization_constant` in standard SPIR-V. Before this patch, this would be represented using Intel's VLA extension. However, as this construct is valid SPIR-V (using specialization constant-length SPIR-V array types), we intend to enable this construct using variables of specialization constant-length arrays.

As the pointer type will differ from the original one, a bitcast is needed. We are not creating 2-dimensional arrays, just a variable of a 1D array type.